### ISSUE TITLE
Gumbel_softmax issue in hard mode 

### ISSUE THREAD

**Issue by wjth07** [pytorch/pytorch#48574]
https://github.com/pytorch/pytorch/blob/bdf360f9f233211b0f84bfb461700d32f69545f7/torch/nn/functional.py#L1549
is there should be:
ret = (y_hard - y_soft).detach() + y_soft ？

cc @albanD @mruberry

**Comment by gchanan** [comment]
I don't think it matters -- y_hard will not have a gradient, so the subtraction shouldn't have a gradient either.

**Comment by gchanan** [comment]
Going to close, if you disagree with the assessment please re-open.
